### PR TITLE
Directorylist fixes

### PR DESCRIPTION
**Pull request by McdonaldSeanp**
This updated DirectoryList and the heat commands to create the correct directory structure in the PA MSI

### CODE REVIEW

**Comment by jcoconnor** [comment]
Nice work
@McdonaldSeanp Pulling this down to continue work on it.
Would prefer if the branch and title referred to actual ticket (RE-6746)

File: lib/vanagon/platform/windows.rb [hunk 1]
```diff
@@ -163,9 +163,22 @@ def generate_nuget_package(project)
...
+        #   dir <directory> - Traverse directory to find all sub-files and directories.
+        #   -ke             - Keep Empty directories
+        #   -cg             - Component Group Name
+        #   -gg             - Generate GUIDS now
+        #   -srd            - Suppress root element generation, we want to reference one of the default root elements
+        #                     INSTALLDIR or APPDATADIR in the directorylist.wxs file, not a newly generated one.
+        #   -sreg           - Suppress registry harvesting.
+        #   -dr             - Root DirectoryRef to point all components to
+        #   -var            - Replace "SourceDir" in the @source attributes of all components with a preprocessor variable
+        application_heat_flags = " -dr INSTALLDIR -v -ke -indent 2 -cg ApplicationComponentGroup -gg -srd -t wix/filter.xslt -sreg -var var.ApplicationSourcePath "
```
Comment (jcoconnor) [inline lib/vanagon/platform/windows.rb]: Minor - but would shorten this to app_

File: lib/vanagon/platform/windows.rb [hunk 2]
```diff
@@ -163,9 +163,22 @@ def generate_nuget_package(project)
...
+        application_source_path = "SourceDir/#{project.settings[:base_dir]}/#{project.settings[:company_id]}/#{project.settings[:product_id]}"
+        appdata_heat_flags = "  -dr APPDATADIR -v -ke -indent 2 -cg ApplicationDataComponentGroup -gg -srd -t wix/filter.xslt -sreg -var var.AppDataSourcePath "
```
Comment (jcoconnor) [inline lib/vanagon/platform/windows.rb]: Suggest you capture the following as common heat flags ```common_heat_flags="-v -ke -indent 2 -gg -srd -t wix/filter.xslt -sreg```

File: lib/vanagon/platform/windows.rb [hunk 3]
```diff
@@ -174,22 +187,17 @@ def generate_msi_package(project)
...
-          #   dir <directory> - Traverse directory to find all sub-files and directories.
-          #   -ke             - Keep Empty directories
-          #   -cg             - Component Group Name
-          #   -gg             - Generate GUIDS now
-          #   -srd            - Suppress root element generation, we want to reference the default root element
-          #                     TARGETDIR in the project.wxs file, not a newly generated one.
-          #   -sreg           - Suppress registry harvesting.
-          "cd $(tempdir); \"$$WIX/bin/heat.exe\" dir SourceDir -v -ke -indent 2 -cg #{cg_name} -gg -srd -t wix/filter.xslt -sreg -out wix/#{project.name}-harvest.wxs",
+          "cd $(tempdir); \"$$WIX/bin/heat.exe\" dir #{application_source_path} #{application_heat_flags} -out wix/#{project.name}-harvest-app.wxs",
+          "cd $(tempdir); \"$$WIX/bin/heat.exe\" dir #{appdata_source_path} #{appdata_heat_flags}  -out wix/#{project.name}-harvest-programdata.wxs",
```
Comment (jcoconnor) [inline lib/vanagon/platform/windows.rb]: :+1: to the reduced verbosity here

File: lib/vanagon/platform/windows.rb [hunk 4]
```diff
@@ -174,22 +187,17 @@ def generate_msi_package(project)
...
           # Apply Candle command to all *.wxs files - generates .wixobj files in wix directory.
           # cygpath conversion is necessary as candle is unable to handle posix path specs
-          "cd $(tempdir)/wix/wixobj; for wix_file in `find $(tempdir)/wix -name \'*.wxs\'`; do \"$$WIX/bin/candle.exe\" #{candle_flags} $$(cygpath -aw $$wix_file) || exit 1; done",
+          # the preprocessor variables AppDataSourcePath and ApplicationSourcePath are required due to the -var input to the heat
+          # runs listed above.
+          "cd $(tempdir)/wix/wixobj; for wix_file in `find $(tempdir)/wix -name \'*.wxs\'`; do \"$$WIX/bin/candle.exe\" #{candle_flags} -dApplicationSourcePath=\"#{application_source_path}\" -dAppDataSourcePath=\"#{appdata_source_path}\" $$(cygpath -aw $$wix_file) || exit 1; done",
```
Comment (jcoconnor) [inline lib/vanagon/platform/windows.rb]: Please put the following into the flags assignments above ```-dApplicationSourcePath=\"#{application_source_path}\" -dAppDataSourcePath=\"#{appdata_source_path}\"```

File: lib/vanagon/platform/windows.rb [hunk 5]
```diff
@@ -174,22 +187,17 @@ def generate_msi_package(project)
...
           # run all wix objects through light to produce the msi
           # the -b flag simply points light to where the SourceDir location is
-          "cd $(tempdir)/wix/wixobj; \"$$WIX/bin/light.exe\" #{light_flags} -b $$(cygpath -aw $(tempdir)) -out $$(cygpath -aw $(workdir)/output/#{target_dir}/#{msi_package_name(project)}) *.wixobj",
+          # -loc is required for the UI localization it points to the actual localization .wxl
+          "cd $(tempdir)/wix/wixobj; \"$$WIX/bin/light.exe\" #{light_flags} -b $$(cygpath -aw $(tempdir)) -loc $$(cygpath -aw $(tempdir)/wix/localization/puppet_en-us.wxl) -out $$(cygpath -aw $(workdir)/output/#{target_dir}/#{msi_package_name(project)}) *.wixobj",
```
Comment (jcoconnor) [inline lib/vanagon/platform/windows.rb]: The ```-loc $$(cygpath -aw $(tempdir)/wix/localization/puppet_en-us.wxl)``` belongs to the UI changes - would prefer if you didn't include it in this change as I have a specific comment related to it in the UI work.

File: resources/windows/wix/directorylist.wxs.erb [hunk 6]
```diff
@@ -1,9 +1,8 @@
 <?xml version='1.0' encoding='windows-1252'?>
 <Wix xmlns='http://schemas.microsoft.com/wix/2006/wi' xmlns:util="http://schemas.microsoft.com/wix/UtilExtension">
-
   <Fragment>
-    <DirectoryRef Id='TARGETDIR'>
-      <%= @platform.generate_wix_dirs(self.get_services) %>
+    <DirectoryRef Id="TARGETDIR">
+      <%= @platform.generate_wix_dir_structure(self.get_services, self) %>
```
Comment (jcoconnor) [inline resources/windows/wix/directorylist.wxs.erb]: @McdonaldSeanp I think you would be better putting in the Wix directly for the top level structure and just using the ```get_services``` to list the service exclusions. My reasoning for this is:
1. Its much clearer to have the main structure written out in Wix rather than derived (the directory structure is complex enough without hiding it further :( )
2. What happens if the product has no services - will this code still run correctly.

Using the following segment would work (picked from change 47f522f57508fd6f2bf1458d89a5266afec1a0d2):
```
     <DirectoryRef Id='TARGETDIR'>
      <Directory Id="<%= settings[:base_dir] %>" >
        <Directory Id="<%= settings[:company_id] %>" Name="<%= settings[:company_name] %>">
          <Directory Id='INSTALLDIR' Name="<%= settings[:product_id] %>" />
        </Directory>
      </Directory>
      <Directory Id="CommonAppDataFolder" Name="CommonAppData">
        <Directory Id="<%= settings[:company_id] %>AppData" Name="<%= settings[:company_id] %>">
        </Directory>
      </Directory>
    </DirectoryRef>
```
This segment also corrects:
1. Correct company name for "Puppet Labs"
2. Uses Id only for the AppData and Program[64]Directory

File: lib/vanagon/platform/windows.rb [hunk 7]
```diff
@@ -289,33 +297,44 @@ def output_dir(target_repo = "")
...
+      # based on that list
+      #
+      # @param services, list of all services in a project
+      # @param project, actual project we are creating the directory structure for
+      def generate_wix_dir_structure(services, project)
+        items = []
+        # All service files will need a directory reference
+        services.map { |svc| items.push({ :path => File.dirname(svc.service_file.gsub(/\\/, "/")), :element_to_add => "<Directory Id=\"#{svc.bindir_id}\" />\n" }) }
+        # INSTALLDIR is the root installation directory for the project
+        items.push({ :path => "SourceDir/#{project.settings[:base_dir]}/#{project.settings[:company_id]}/#{project.settings[:product_id]}", :element_to_add => "<Directory Id=\"INSTALLDIR\" />\n" })
```
Comment (jcoconnor) [inline lib/vanagon/platform/windows.rb]: ```:company_id``` should be ```:company_name```

You also need to make sure that the ```#{project.settings[:base_dir]}``` does not use the ```"Name"``` field as this cause the ```ProgamFiles64Folder``` directory to be created.

I think it would be better to actually express these directories directly in the Wix - see https://github.com/puppetlabs/vanagon/pull/318/files#r56167505 below

File: lib/vanagon/platform/windows.rb [hunk 8]
```diff
@@ -289,33 +297,44 @@ def output_dir(target_repo = "")
...
+        # APPDATADIR is the root directory for all application data (config files etc.)
+        items.push({ :path => "SourceDir/CommonAppDataFolder", :element_to_add => "<Directory Id=\"APPDATADIR\" />\n" })
```
Comment (jcoconnor) [inline lib/vanagon/platform/windows.rb]: You need to make sure that ```CommonAppDataFolder``` does the ```"Name"``` field as as above - again see my suggestion in https://github.com/puppetlabs/vanagon/pull/318/files#r56167505 below

File: resources/windows/wix/project.wxs.erb [hunk 9]
```diff
@@ -17,7 +17,10 @@
       Compressed="yes"
       Platform="<%= @platform.architecture %>" />
 
-      <!--  We are discussing a MAINT PR to move the componentfilegroup to seperate file
```
Comment (jcoconnor) [inline resources/windows/wix/project.wxs.erb]: The more I look at this file I am convinced that we do need to create a ```componentgroups.wxs.erb``` file, so either:
1. We leave this comment in or
2. Ticket a task for it (1 pointer)

Have created ticket: [RE-6757](https://tickets.puppetlabs.com/browse/RE-6757)

File: resources/windows/wix/project.wxs.erb [hunk 10]
```diff
@@ -25,7 +28,8 @@
 
     <Feature Id="<%= settings[:product_id] %>Runtime" Title="<%= settings[:product_id] %> Runtime" Level="1">
       <!-- We can add all components by referencing this one thing -->
-      <ComponentGroupRef Id="ProductComponentGroup" />
+      <ComponentGroupRef Id="ApplicationComponentGroup" />
```
Comment (jcoconnor) [inline resources/windows/wix/project.wxs.erb]: Again wondering about shorter names e.g. ```AppCompGroup``` - not essential - just a thought.

**Comment by McdonaldSeanp** [comment]
This has been tested and works with the current PA changes.

**Comment by jcoconnor** [comment]
@McdonaldSeanp I have run a test on this.
Program Files looks fine, but I'm not sure the Application Data directories are quite correct:

1. C:\ProgramData\PuppetLabs is still being created - this appeared to happen after the files installed, so it might an issue with the service startup scripts.
2. Also is "C:\ProgramData\Puppet Labs\PuppetLabs" correct ?

**Comment by jcoconnor** [comment]
@McdonaldSeanp I checked https://github.com/melissa/puppet-specifications/blob/maint/master/windows-path-updates/file_paths.md

The Program Data directory should be: ```C:\ProgramData\PuppetLabs```  (no space).
So apologies if I misdirected you earlier on this.

**Comment by McdonaldSeanp** [comment]
The above has been remedied and tested, this is ready for merge

**Comment by jcoconnor** [comment]
:+1:

File: lib/vanagon/platform/windows.rb [hunk 11]
```diff
@@ -174,21 +189,16 @@ def generate_msi_package(project)
...
-          #   dir <directory> - Traverse directory to find all sub-files and directories.
-          #   -ke             - Keep Empty directories
-          #   -cg             - Component Group Name
-          #   -gg             - Generate GUIDS now
-          #   -srd            - Suppress root element generation, we want to reference the default root element
-          #                     TARGETDIR in the project.wxs file, not a newly generated one.
-          #   -sreg           - Suppress registry harvesting.
-          "cd $(tempdir); \"$$WIX/bin/heat.exe\" dir SourceDir -v -ke -indent 2 -cg #{cg_name} -gg -srd -t wix/filter.xslt -sreg -out wix/#{project.name}-harvest.wxs",
+          "cd $(tempdir); \"$$WIX/bin/heat.exe\" dir #{app_source_path} #{app_heat_flags} -out wix/#{project.name}-harvest-app.wxs",
+          "cd $(tempdir); \"$$WIX/bin/heat.exe\" dir #{appdata_source_path} #{appdata_heat_flags}  -out wix/#{project.name}-harvest-programdata.wxs",
```
Comment (melissa) [inline lib/vanagon/platform/windows.rb]: Is it possible a project wont install anything into `appdata_source_path`? If so, what will this code do?

Comment (jcoconnor) [inline lib/vanagon/platform/windows.rb]: That could be the case.
Would suggest we ticket this and deal with after end of March.
Puppet-Agent will be using both, so we are ok for the moment.

File: lib/vanagon/platform/windows.rb [hunk 12]
```diff
@@ -163,9 +163,24 @@ def generate_nuget_package(project)
...
+        #   -srd            - Suppress root element generation, we want to reference one of the default root elements
+        #                     INSTALLDIR or APPDATADIR in the directorylist.wxs file, not a newly generated one.
+        #   -sreg           - Suppress registry harvesting.
+        #   -dr             - Root DirectoryRef to point all components to
+        #   -var            - Replace "SourceDir" in the @source attributes of all components with a preprocessor variable
+        app_heat_flags = " -dr INSTALLDIR -v -ke -indent 2 -cg AppComponentGroup -gg -srd -t wix/filter.xslt -sreg -var var.AppSourcePath "
+        app_source_path = "SourceDir/#{project.settings[:base_dir]}/#{project.settings[:company_id]}/#{project.settings[:product_id]}"
+        appdata_heat_flags = "  -dr APPDATADIR -v -ke -indent 2 -cg AppDataComponentGroup -gg -srd -t wix/filter.xslt -sreg -var var.AppDataSourcePath "
+        appdata_source_path = "SourceDir/CommonAppDataFolder"
+        # Candle.exe preprocessor vars are required due to the above double run of
```
Comment (mckern) [inline lib/vanagon/platform/windows.rb]: Double run of what?

File: lib/vanagon/platform/windows.rb (hunk 11 above)
Comment (mckern) [inline lib/vanagon/platform/windows.rb]: @jcoconnor Can you take point on getting that ticket opened up? What would the side effects be if the location under `appdata_source_path` was empty?

Comment (McdonaldSeanp) [inline lib/vanagon/platform/windows.rb]: @melissa It's likely it would make what is essentially an empty WiX fragment, but again, as John said, this is ticketed for after we finished

File: lib/vanagon/platform/windows.rb (hunk 12 above)
Comment (McdonaldSeanp) [inline lib/vanagon/platform/windows.rb]: Oh, ooops

File: lib/vanagon/platform/windows.rb [hunk 13]
```diff
@@ -289,33 +299,46 @@ def output_dir(target_repo = "")
...
+      # @param project, actual project we are creating the directory structure for
+      def generate_service_bin_dirs(services, project)
+        items = []
+        # All service files will need a directory reference
+        services.map do |svc|
+          items.push({
+            :path => strip_and_format_path(svc.service_file, project),
+            :element_to_add => "<Directory Id=\"#{svc.bindir_id}\" />\n"
+          })
+        end
```
Comment (mckern) [inline lib/vanagon/platform/windows.rb]: I think you can lose the empty array initialization here?

```ruby
items = services.map do |svc|
  {
    :path => strip_and_format_path(svc.service_file, project),
    :element_to_add => "<Directory Id=\"#{svc.bindir_id}\" />\n"
  }
end
```

File: lib/vanagon/platform/windows.rb (hunk 11 above)
Comment (jcoconnor) [inline lib/vanagon/platform/windows.rb]: Its not ticketed yet, but I'll make a point of doing it tomorrow morning.

File: lib/vanagon/platform/windows.rb [hunk 14]
```diff
@@ -289,33 +299,46 @@ def output_dir(target_repo = "")
...
         # an n-ary tree structure matching the specs from the input
-        directories.each do |dir|
+        items.each do |item|
           # Always start at the beginning
           curr = root
-          names = strip_path(dir[:path])
-          # The last entry in this list will be the actual file,
-          # which we do not want, we only want it's base path
-          names.pop
+          names = item[:path].split(/\//)
```
Comment (mckern) [inline lib/vanagon/platform/windows.rb]: I'd lose the regex here. Consider:

```ruby
item[:path].split(File::SEPARATOR) # (Array)
```

or 

```ruby
Pathname.new(item[:path]).each_filename.to_a # (Array)
```

File: lib/vanagon/platform/windows.rb (hunk 11 above)
Comment (McdonaldSeanp) [inline lib/vanagon/platform/windows.rb]: @mckern it's actually not possible for CommonAppDataFolder to not exist, it's a windows default.

File: lib/vanagon/platform/windows.rb [hunk 15]
```diff
@@ -289,33 +299,46 @@ def output_dir(target_repo = "")
...
           names.each do |name|
             curr = insert_child(curr, name)
           end
           # at this point, curr will be the top dir, override the id if
           # id exists
-          curr[:bindir_ids].push(dir[:bindir_id])
+          curr[:elements_to_add].push(item[:element_to_add])
         end
         return generate_wix_from_graph(root)
```
Comment (mckern) [inline lib/vanagon/platform/windows.rb]: I'd lose the explicit `return` on a lot of these methods; that's useful when you're exiting early because the work in a method is already done, but if it's the end of the method you should just call/state the value you want to return. In this case, that's `generate_wix_from_graph(root)` without the explicit `return`.

File: lib/vanagon/platform/windows.rb [hunk 16]
```diff
@@ -339,21 +362,14 @@ def insert_child(curr, name)
...
-          # roots on windows
-          if path.start_with?("/") || path.start_with?("\\") || path.start_with?("SourceDir") || path =~ (/([A-Za-z])*\:(\/|\\)/)
-            path = path.sub(/\/|\\|([A-Za-z])*\:(\/|\\)|(\/|\\)?(SourceDir)(\/|\\)?/, '')
-          end
-          names = path.split(/\/|\\/)
-        end
-        return names
+      # @param [@project] project object
+      def strip_and_format_path(path, project)
+        path_regex = /(\\|\/)?SourceDir(\\|\/)#{project.settings[:base_dir]}(\\|\/)#{project.settings[:company_id]}(\\|\/)#{project.settings[:product_id]}(\\|\/)/
```
Comment (mckern) [inline lib/vanagon/platform/windows.rb]: ... that's some regex you got there.

File: lib/vanagon/platform/windows.rb (hunk 11 above)
Comment (melissa) [inline lib/vanagon/platform/windows.rb]: Cool, can you link the ticket here so we know it's been made?

File: lib/vanagon/platform/windows.rb [hunk 17]
```diff
@@ -163,9 +163,24 @@ def generate_nuget_package(project)
...
+        #   -ke             - Keep Empty directories
+        #   -cg             - Component Group Name
+        #   -gg             - Generate GUIDS now
+        #   -srd            - Suppress root element generation, we want to reference one of the default root elements
+        #                     INSTALLDIR or APPDATADIR in the directorylist.wxs file, not a newly generated one.
+        #   -sreg           - Suppress registry harvesting.
+        #   -dr             - Root DirectoryRef to point all components to
+        #   -var            - Replace "SourceDir" in the @source attributes of all components with a preprocessor variable
+        app_heat_flags = " -dr INSTALLDIR -v -ke -indent 2 -cg AppComponentGroup -gg -srd -t wix/filter.xslt -sreg -var var.AppSourcePath "
+        app_source_path = "SourceDir/#{project.settings[:base_dir]}/#{project.settings[:company_id]}/#{project.settings[:product_id]}"
```
Comment (melissa) [inline lib/vanagon/platform/windows.rb]: Part of me wants to put checks in to make sure the user has at least these three variables set, so the build fails early. That's not a requirement for this PR though

File: lib/vanagon/platform/windows.rb [hunk 18]
```diff
@@ -163,9 +163,24 @@ def generate_nuget_package(project)
...
+        #   dir <directory> - Traverse directory to find all sub-files and directories.
+        #   -ke             - Keep Empty directories
+        #   -cg             - Component Group Name
+        #   -gg             - Generate GUIDS now
+        #   -srd            - Suppress root element generation, we want to reference one of the default root elements
+        #                     INSTALLDIR or APPDATADIR in the directorylist.wxs file, not a newly generated one.
+        #   -sreg           - Suppress registry harvesting.
+        #   -dr             - Root DirectoryRef to point all components to
+        #   -var            - Replace "SourceDir" in the @source attributes of all components with a preprocessor variable
+        app_heat_flags = " -dr INSTALLDIR -v -ke -indent 2 -cg AppComponentGroup -gg -srd -t wix/filter.xslt -sreg -var var.AppSourcePath "
```
Comment (melissa) [inline lib/vanagon/platform/windows.rb]: Where is `var.AppSourcePath` getting set? Is that something provided by wix by default?

Comment (jcoconnor) [inline lib/vanagon/platform/windows.rb]: Thats a directive to heat to use that variable as a reference instead of source dir
If you look down, you will see that we actually give it a value for the candle run.

Comment (McdonaldSeanp) [inline lib/vanagon/platform/windows.rb]: Yep, var.AppSourcePath is a preprocessor variable we pass to candle when it compiles these files. refer to: https://github.com/McdonaldSeanp/vanagon/blob/feature/RE-6470/allow_config_files_to_stay/lib/vanagon/platform/windows.rb#L182

File: lib/vanagon/platform/windows.rb (hunk 17 above)
Comment (McdonaldSeanp) [inline lib/vanagon/platform/windows.rb]: Yep, these deps are used outside this, so that would be its own PR

**Comment by McdonaldSeanp** [comment]
Retested with all the new changes. Still builds an MSI, still installs correctly.

File: resources/windows/wix/filter.xslt.erb [hunk 19]
```diff
@@ -13,9 +13,15 @@
...
+  <!-- we have to substitue the SourceDir\\etc. etc. for $(var.ProjectSourcePath) in order to facilitate heat -->
+  <!--   being run in two different places, specifically it no longer runs from SourceDir so we needed to pass -->
+  <!--   in ProjectSourcePath to the heat runs. Thus when this filter does its work the source attribute will -->
+  <!--   look like $(var.ProjectSourcePath)/puppet/bin/rubyw.exe for example, not  -->
+  <!--   SourceDir/Program64FilesFolder/Puppet/puppet/bin/ruby.exe -->
   <%- service_files = Array.new -%>
   <%- get_services.each do |service| -%>
-    <%- service_files << service.service_file -%>
+
+    <%- service_files << service.service_file.sub("SourceDir\\#{self.settings[:base_dir]}\\#{self.settings[:company_id]}\\#{self.settings[:product_id]}", "$(var.AppSourcePath)") -%>
```
Comment (melissa) [inline resources/windows/wix/filter.xslt.erb]: Is this a change we should force in the service files rather than here?

Comment (melissa) [inline resources/windows/wix/filter.xslt.erb]: Especially because we're manually specifying these paths anyway

Comment (McdonaldSeanp) [inline resources/windows/wix/filter.xslt.erb]: We can't, I can explain this in way better detail offline of these comments (because it takes forever to explain) but there is a strict restriction that this stays here.

File: resources/windows/wix/project.wxs.erb [hunk 20]
```diff
@@ -17,7 +17,10 @@
       Compressed="yes"
       Platform="<%= @platform.architecture %>" />
 
-      <!--  We are discussing a MAINT PR to move the componentfilegroup to seperate file 
+    <UI>
+      <UIRef Id="WixUI_PuppetInstallDir"/>
```
Comment (melissa) [inline resources/windows/wix/project.wxs.erb]: Should this ID be more generic?

File: resources/windows/wix/filter.xslt.erb (hunk 19 above)
Comment (McdonaldSeanp) [inline resources/windows/wix/filter.xslt.erb]: Suffice it to say the service files need to point to the right place, but the wix heat runs also need to point to the right place, and because of that this needs to go here

**Comment by mckern** [comment]
There's some Ruby in here that I think could use a little polish but not enough to block on. :+1: for the Ruby; I can't speak to the WiX components.

File: resources/windows/wix/project.wxs.erb (hunk 20 above)
Comment (McdonaldSeanp) [inline resources/windows/wix/project.wxs.erb]: This shouldn't be in this PR, oops

**Comment by melissa** [comment]
I'm hesitant to merge this until after the ticket for `appdata_source_path` has been created...

File: resources/windows/wix/filter.xslt.erb (hunk 19 above)
Comment (McdonaldSeanp) [inline resources/windows/wix/filter.xslt.erb]: Actually I can explain this here. when this file filters out things in the heat runs, it's not actually anything but an xml filter, so it knows nothing about WiX dirs or anything. All it knows is to filter out the component if it's Source attribute looks like what we are matching. So the problem is, at the point in which the filter is looking for the file, candle has not run yet, so none of the preprocessor variables have been replaced, Thus we have to look for the string of the actual preprocessor variable, not what it will resolve to.

**Comment by jcoconnor** [comment]
@melissa and @mckern Ticket [RE-6761](https://tickets.puppetlabs.com/browse/RE-6766) created for this.
